django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -0,0 +1,493 @@
+        """
+        Slice queryset based on the requested page.
+        """
+        paginator = CachedPaginator(
+            queryset,
+            self.page_size,
+            cache_key=self._get_paginator_cache_key(),
+            cache_vary=self._get_full_cache_vary(params),
+            cache_bust_condition=CacheBustCondition.any_package_updated,
+        )
+
+        # PageNotAnInteger error won't be raised here since deserializer
+        # has already checked that the page parameter is an integer.
+        try:
+            page = paginator.page(params["page"])
+        except EmptyPage:
+            raise ParseError("Page index error: no results on requested page")
+
+        return page
+
+    def _serialize_results(
+        self,
+        params: OrderedDict,
+        package_page: Page,
+    ) -> PackageListResponseSerializer:
+        """
+        Format results to transportation.
+        """
+        packages = [
+            {
+                "categories": p.community_listings.all()[0].categories.all(),
+                "community_identifier": p.community_listings.all()[
+                    0
+                ].community.identifier,
+                "description": p.latest.description,
+                "download_count": p.download_count,
+                "icon_url": p.latest.icon.url if bool(p.latest.icon) else None,
+                "is_deprecated": p.is_deprecated,
+                "is_nsfw": p.community_listings.all()[0].has_nsfw_content,
+                "is_pinned": p.is_pinned,
+                "last_updated": p.date_updated,
+                "namespace": p.namespace.name,
+                "name": p.name,
+                "rating_count": p.rating_count,
+                "size": p.latest.file_size,
+            }
+            for p in package_page.object_list
+        ]
+
+        (prev_url, next_url) = self._get_sibling_pages(params, package_page)
+
+        return PackageListResponseSerializer(
+            {
+                "count": package_page.paginator.count,
+                "previous": prev_url,
+                "next": next_url,
+                "results": packages,
+            },
+        )
+
+    def _get_full_cache_vary(self, params: OrderedDict) -> str:
+        """
+        Return cache key for CachedPaginator.
+        """
+        cache_vary = self._get_paginator_cache_vary_prefix()
+        cache_vary += f".{params['deprecated']}"
+        cache_vary += f".{params['nsfw']}"
+        cache_vary += f".{sorted(params['included_categories'])}"
+        cache_vary += f".{sorted(params['excluded_categories'])}"
+        cache_vary += f".{params.get('section', '-')}"
+        cache_vary += f".{params.get('q', '-')}"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{self.page_size}"
+        return cache_vary
+
+    def _get_sibling_pages(
+        self,
+        params: OrderedDict,
```

Comment:
Not sure I follow the logic. If the serializer is passed, wouldn't I have to validate the params separately in each function where the values are accessed, or alternatively jump around to see if the validation is already called earlier?

Reply:
`validated_data` isn't available before validation I'm quite sure

Reply:
So I need to use `hasattr` or try-catch in every function that wants to access the data? I'm going to skip this suggestion.

Reply:
How about just accessing it as you do now, but rather than it being any arbitrary dictionary, it has at least implied types associated with it when done through the serializer? That was the main focus of my feedback. In both cases we just assume the data to exist before accessing it.

Reply:
I've no idea what you're talking about, perhaps you could provide a code sample showing what you mean.